### PR TITLE
syncer: add a struct `queryEventContext` to manage variables

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve? 
Part of https://github.com/pingcap/dm/issues/2064

### What is changed and how it works?

- add a new struct `queryEventContext` to wrap all variables handleQueryEvent used.

```go
// May be modified in the future
type queryEventContext struct {
	eventContext

	p         *parser.Parser // used parser
	ddlSchema string         // used schema
	originSQL string         // before split
	// split multi-schema change DDL into multiple one schema change DDL due to TiDB's limitation
	splitedDDLs []string // after split before online ddl
	appliedDDLs []string // after onlineDDL apply if onlineDDL != nil and track, before route
	needHandleDDLs  []string // after route

	ddlInfo             *shardingDDLInfo
	needTrackDDLs       []trackedDDL
	sourceTbls          map[string]map[string]struct{} // db name -> tb name
	onlineDDLTableNames map[string]*filter.Table
}
```

### Check List 

Tests 

 - No code

Code changes

Side effects

Related changes

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -2334,88 +2351,98 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
 		return nil
 	}
 
-	usedSchema := string(ev.Schema)
-	parser2, err := event.GetParserForStatusVars(ev.StatusVars)
+	qec := &queryEventContext{
+		eventContext:    &ec,
+		ddlSchema:       string(ev.Schema),
+		originSQL:       utils.TrimCtrlChars(originSQL), // Will TrimCtrlChars affect code?
```
Comment (Ehco1996) [inline syncer/syncer.go:2357]: i think `TrimCtrlChars` doesn't  affect originSQL, it just trims invalid char from the sql string, but i am wondering why we need this func to handle sql in binlog? need some suggestions from other reviewer

Comment (Ehco1996) [reply syncer/syncer.go:2357]: but if we trim sql here, we don't need to trim it again in `parseDDLSQL`

File: syncer/syncer.go [hunk 2]
```diff
@@ -2579,7 +2600,7 @@ func (s *Syncer) handleQueryEventNoSharding(
 		}
 	})
 
-	job := newDDLJob(nil, needHandleDDLs, *ec.lastLocation, *ec.startLocation, *ec.currentLocation, sourceTbls, originSQL, ec.header)
+	job := newDDLJob(nil, qec.needHandleDDLs, *qec.lastLocation, *qec.startLocation, *qec.currentLocation, qec.sourceTbls, qec.originSQL, qec.header)
```
Comment (Ehco1996) [inline syncer/syncer.go:2603]: `newDDLJob ` could also use `qec` as arg

File: syncer/syncer.go [hunk 3]
```diff
@@ -2334,88 +2351,98 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
...
+	qec.p, err = event.GetParserForStatusVars(ev.StatusVars)
 	if err != nil {
 		log.L().Warn("found error when get sql_mode from binlog status_vars", zap.Error(err))
 	}
 
-	parseResult, err := s.parseDDLSQL(originSQL, parser2, usedSchema)
+	parseResult, err := s.parseDDLSQL(originSQL, qec.p, qec.ddlSchema)
 	if err != nil {
-		ec.tctx.L().Error("fail to parse statement", zap.String("event", "query"), zap.String("statement", originSQL), zap.String("schema", usedSchema), zap.Stringer("last location", ec.lastLocation), log.WrapStringerField("location", ec.currentLocation), log.ShortError(err))
+		qec.tctx.L().Error("fail to parse statement", zap.String("event", "query"), zap.String("statement", originSQL), zap.String("schema", qec.ddlSchema), zap.Stringer("last location", qec.lastLocation), log.WrapStringerField("location", qec.currentLocation), log.ShortError(err))
```
[the remark this answers is not in the document]
Comment (okJiang) [reply syncer/syncer.go:2371]: I think so too. We can create a `queryEventContext.String()`, and log `zap.Stringer(qec)`.

File: syncer/job.go [hunk 4]
```diff
@@ -138,6 +133,10 @@ func newDDLJob(ddlInfo *shardingDDLInfo, ddls []string, location, startLocation,
 			}
 		}
 		j.sourceTbl = sourceTbl
+	} else {
+		j.sourceTbl = map[string][]string{qec.ddlInfo.tables[0][0].Schema: {qec.ddlInfo.tables[0][0].Name}}
```
Comment (Ehco1996) [inline syncer/job.go:137]: please check `qec.ddlinfo` is nil ,or  this may cause panic 😂

File: syncer/job.go [hunk 5]
```diff
@@ -109,27 +109,22 @@ func newDMLJob(tp opType, sourceSchema, sourceTable, targetSchema, targetTable,
 // newDDL job is used to create a new ddl job
 // when cfg.ShardMode == "", ddlInfo == nil，sourceTbls != nil, we use sourceTbls to record ddl affected tables.
 // when cfg.ShardMode == ShardOptimistic || ShardPessimistic, ddlInfo != nil, sourceTbls == nil.
-func newDDLJob(ddlInfo *shardingDDLInfo, ddls []string, location, startLocation, cmdLocation binlog.Location,
-	sourceTbls map[string]map[string]struct{}, originSQL string, eventHeader *replication.EventHeader) *job {
+func newDDLJob(qec *queryEventContext) *job {
```
Comment (Ehco1996) [inline syncer/job.go:112]: also need to change the comment of `newDDLJob`

File: syncer/job.go [hunk 6]
```diff
@@ -109,27 +109,22 @@ func newDMLJob(tp opType, sourceSchema, sourceTable, targetSchema, targetTable,
...
 	}
 
-	if ddlInfo != nil {
-		j.sourceTbl = map[string][]string{ddlInfo.tables[0][0].Schema: {ddlInfo.tables[0][0].Name}}
-		j.targetSchema = ddlInfo.tables[1][0].Schema
-		j.targetTable = ddlInfo.tables[1][0].Name
-	} else if sourceTbls != nil {
-		sourceTbl := make(map[string][]string, len(sourceTbls))
-		for schema, tbMap := range sourceTbls {
+	if qec.sourceTbls != nil && len(qec.sourceTbls) != 0 {
```
Comment (Ehco1996) [inline syncer/job.go:125]: seems dont't need this check? `&& len(qec.sourceTbls) != 0` , the for loop in L127 is already do this

File: syncer/syncer.go [hunk 7]
```diff
@@ -2319,7 +2319,46 @@ func (s *Syncer) handleRowsEvent(ev *replication.RowsEvent, ec eventContext) err
...
+		lastLocation = qec.lastLocation.String()
+	}
+	var needHandleDDLs, shardingReSync string
+	if qec.needHandleDDLs != nil {
+		needHandleDDLs = strings.Join(qec.needHandleDDLs, ",")
+	}
+	if qec.shardingReSync != nil {
+		shardingReSync = qec.shardingReSync.String()
+	}
+	return fmt.Sprintf("{schema: %s, originSQL: %s, startLocation: %s, currentLocation: %s, lastLocation: %s, re-sync: %s, needHandleDDLs: %s}",
```
Comment (Ehco1996) [inline syncer/syncer.go:2357]: If we want to splice a large number of strings, we can use `string buffer` for better performance
ref  https://stackoverflow.com/questions/1760757/how-to-efficiently-concatenate-strings-in-go

**Review by Ehco1996 — COMMENTED** [review]
rest LGTM ， Great job 🎉 This PR will make "handleQueryEvent" easier to read.

File: syncer/syncer.go (hunk 7 above)
Comment (okJiang) [reply syncer/syncer.go:2357]: maybe `strings.Join` is ok. https://segmentfault.com/a/1190000040275250

File: syncer/syncer.go (hunk 1 above)
Comment (lance6716) [reply syncer/syncer.go:2375]: pingcap/tidb-enterprise-tools#258

File: syncer/syncer.go [hunk 8]
```diff
@@ -2330,88 +2369,98 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
...
 
-	ec.tctx.L().Info("", zap.String("event", "query"), zap.String("statement", originSQL), zap.String("schema", usedSchema), zap.Stringer("last location", ec.lastLocation), log.WrapStringerField("location", ec.currentLocation))
-	*ec.lastLocation = *ec.currentLocation // update lastLocation, because we have checked `isDDL`
+	qec.tctx.L().Info("", zap.String("event", "query"), zap.Stringer("queryEventContext", qec))
+	*qec.lastLocation = *qec.currentLocation // update lastLocation, because we have checked `isDDL`
 
 	// TiDB can't handle multi schema change DDL, so we split it here.
 	// for DDL, we don't apply operator until we try to execute it. so can handle sharding cases
 	// We use default parser because inside function where need parser, sqls are came from parserpkg.SplitDDL, which is StringSingleQuotes, KeyWordUppercase and NameBackQuotes
 	// TODO: save stmt, tableName to avoid parse the sql to get them again
```
Comment (lance6716) [inline syncer/syncer.go:2450]: can we resolve some part of this TODO?

File: syncer/syncer.go [hunk 9]
```diff
@@ -2467,47 +2513,49 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
...
 		}
 
-		needHandleDDLs = append(needHandleDDLs, sqlDDL)
-		needTrackDDLs = append(needTrackDDLs, trackedDDL{rawSQL: sql, stmt: stmt, tables: tables})
+		qec.needHandleDDLs = append(qec.needHandleDDLs, sqlDDL)
+		qec.needTrackDDLs = append(qec.needTrackDDLs, trackedDDL{rawSQL: sql, stmt: stmt, tables: tables})
 		// TODO: current table checkpoints will be deleted in track ddls, but created and updated in flush checkpoints,
 		//       we should use a better mechanism to combine these operations
-		recordSourceTbls(sourceTbls, stmt, tables[0][0])
+		if !s.cfg.IsSharding {
```
Comment (lance6716) [inline syncer/syncer.go:2550]: `IsSharding` should be deprecated, we now use ShardMode more often

File: syncer/syncer.go (hunk 8 above)
Comment (okJiang) [reply syncer/syncer.go:2450]: We don't resolve this TODO this pr. Maybe will resolve it in https://github.com/pingcap/dm/pull/2090.

**Review by Ehco1996 — APPROVED** [review]
LGTM

File: syncer/syncer.go (hunk 7 above)
Comment (lichunzhu) [reply syncer/syncer.go:2357]: Why don't we use "+" or builder to do this?

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: syncer/syncer.go [hunk 10]
```diff
@@ -2330,88 +2369,98 @@ func (s *Syncer) handleQueryEvent(ev *replication.QueryEvent, ec eventContext, o
...
+	qec.tctx.L().Info("", zap.String("event", "query"), zap.Stringer("queryEventContext", qec))
+	*qec.lastLocation = *qec.currentLocation // update lastLocation, because we have checked `isDDL`
 
 	// TiDB can't handle multi schema change DDL, so we split it here.
 	// for DDL, we don't apply operator until we try to execute it. so can handle sharding cases
 	// We use default parser because inside function where need parser, sqls are came from parserpkg.SplitDDL, which is StringSingleQuotes, KeyWordUppercase and NameBackQuotes
 	// TODO: save stmt, tableName to avoid parse the sql to get them again
-	sqls, onlineDDLTableNames, err := s.splitAndFilterDDL(ec, parser.New(), parseResult.stmt, usedSchema)
+	qec.p = parser.New()
+	qec.appliedDDLs, qec.onlineDDLTables, err = s.splitAndFilterDDL(*qec.eventContext, qec.p, parseResult.stmt, qec.ddlSchema)
```
[the remark this answers is not in the document]
Comment (okJiang) [reply syncer/syncer.go:2452]: We will change it in https://github.com/pingcap/dm/pull/2090.

File: syncer/syncer.go (hunk 7 above)
Comment (okJiang) [reply syncer/syncer.go:2357]: > Why don't we use "+" or builder to do this?

The implementation of `join` is the `builder`

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**